### ISSUE TITLE
Database.SetHead() doesn't do auto-merging of concurrency dataset changes

### ISSUE THREAD

**Issue by aboodman** [attic-labs/noms#3280]
Database.Commit() "auto-merges" concurrent changes to different dataset heads. IOW, if you call:

`db.CommitValue(ds1, newVal)`

... and `db` has moved forward in another process, but in that new database version, `ds1` still has the same head as the value you specified, then we allow the commit silently.

But Database.SetHead() does not have this same behavior. Seems like it should. Also, I wonder about the other APIs of Database that mutate dataset heads -- FastForward, Delete.

**Comment by aboodman** [comment]
@cmasone-attic, @rafael-atticlabs thoughts?

**Comment by cmasone-attic** [comment]
For SetHead and Delete...I think that'd be ok. Those are sorta meant to be
"destructive"

For FastForward, I kinda want to think about it for a minute

**Comment by rafael-atticlabs** [comment]
Yeah. Makes sense. In general, the unit of atomicity is a dataset. It might be possible, but it's certainly not explicitly supported to do anything more. Allowing dataset heads to move independently seems consistent.

**Comment by cmasone-attic** [comment]
Reading the comment for `FastForward`, I'd say that we shouldn't add this kind of behavior to it. It specifically allows moving the head forward only in the case that the proposed head is a descendant of the existing head.

**Comment by aboodman** [comment]
That's the dataset head, not the database head (e.g., root), right?

The behavior I'm advocating is that the general algorithm for all
dataset-modifying functions is:

1. Attempt to perform operation on local (in-memory) view of db
2. If it can't be performed on local db, fail
3. Commit to db
4. If commit to db fails because root of db has moved forward, go back to 1
and try again with new db root

So FF would still fail if we ever find that dataset head has moved in
another process. But we'd keep trying to FF as long as only OTHER dataset
heads have moved forward in other processes.

**Comment by cmasone-attic** [comment]
Ah, I misread. Yeah, that seems like it makes sense
